New article - Migrate from Newtonsoft.Json (#16225)
dotnet/docs: docs/standard/serialization/system-text-json-converters-how-to.md

* To override the default behavior of a built-in converter. For example, you might want `DateTime` values to be represented by mm/dd/yyyy format instead of the default  ISO 8601-1:2019 format.
* To support a custom value type. For example, a `PhoneNumber` struct.

> 

Author: They were moved to the Migrate doc.  I could keep them here but I thought that moving them would be better than duplicating them in the Migrate doc or linking to here from the Migrate doc. I could bring them back if you or @steveharter prefer to see them here.

* To override the default behavior of a built-in converter. For example, you might want `DateTime` values to be represented by mm/dd/yyyy format instead of the default  ISO 8601-1:2019 format.
* To support a custom value type. For example, a `PhoneNumber` struct.

> You can also write custom converters to customize or extend `System.Text.Json` with functionality not included in the current release. The following scenarios are covered later in this article:

Author: It's used everywhere, so I thought making every occurrence a link would look too busy and I made it an xref link only on first occurrence. Do you prefer links on every occurrence?  Or maybe an intermediate approach such as first occurrence in each section?

> You can also write custom converters to customize or extend `System.Text.Json` with functionality not included in the current release. The following scenarios are covered later in this article:

Reviewer: I think I missed the first occurrence.

Generally, we link on the first occurrence. This article might be long enough (and you have a lot of links to jump to specific sections) that an xref in the first occurrence each section would be the best choice here.

> You can also write custom converters to customize or extend `System.Text.Json` with functionality not included in the current release. The following scenarios are covered later in this article:

Author: Ah, that's because first occurrence I linked to the overview doc.  I'll change it to xref and implement the first-per-section approach.

* Create a class that derives from <xref:System.Text.Json.Serialization.JsonConverter%601> where `T` is the type to be serialized and deserialized.
* Override the `Read` method to deserialize the incoming JSON and convert it to type `T`. Use the <xref:System.Text.Json.Utf8JsonReader> that is passed to the method to read the JSON.
* Override the `Write` method to serialize the incoming object of type `T`. Use the <xref:System.Text.Json.Utf8JsonWriter> that is passed to the method to write the JSON.
* Override the `CanConvert` method only if necessary. The default implementation returns `true` when the type to convert is of type `T`. Therefore, converters that support only type `T` don't need to override this method. For an example of a converter that does need to override this method, see the [polymorphic deserialization](#support-polymorphic-deserialization) section later in this article.

> You can refer to the [built-in converters source code](https://github.com/dotnet/runtime/tree/f77914d4c9045a01b3a91b63c28805f24850c274/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/) as reference implementations for writing custom converters.

Author: I originally had linked to master, but Pranav and Ahson recommended a specific commit so the links wouldn't break if code is moved around or changed. Do we have a team standard for links to source?

> You can refer to the [built-in converters source code](https://github.com/dotnet/runtime/tree/f77914d4c9045a01b3a91b63c28805f24850c274/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/) as reference implementations for writing custom converters.

Reviewer: We don't have a standard. I think I like this approach, because it is 'fixed'. I wanted to make sure it was an explicit choice.

A built-in converter is chosen only if no applicable custom converter is registered.

## Other custom converter samples

> The [Migrate from Newtonsoft.Json to System.Text.Json](system-text-json-migrate-from-newtonsoft-how-to.md) article contains samples of custom converters.

Author: On second thought regarding this, I'll go ahead and restore them to this doc, and link to them from the Migrate doc.

The [unit tests folder](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Text.Json/tests/Serialization/) in the `System.Text.Json.Serialization` source code includes other custom converter samples, such as:

* `Int32` converter that converts null to 0 on deserialize
* `Int32` converter that allows both string and number values on deserialize
* `Enum` converter
* `List<T>` converter that accepts external data
* `Long[]` converter that works with a comma-delimited list of numbers 

If you need to make a converter that modifies the behavior of an existing built-in converter, you can get [the source code of the existing converter](https://github.com/dotnet/runtime/tree/master/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters) to serve as a starting point for customization.

## Additional resources

* [GitHub issue related to custom converters](https://github.com/dotnet/corefx/issues/36639)

> * [Source code for built-in converters](https://github.com/dotnet/runtime/tree/master/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/)

Author: @steveharter can correct me if I'm wrong, but I think people will want to look at the built-in converters as samples since they can't make derived classes based on them.

> * [Source code for built-in converters](https://github.com/dotnet/runtime/tree/master/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/)

Reviewer 2: Fair point. I am fine with keeping this link. I would update the link to be a fixed commit hash though:
https://github.com/dotnet/runtime/tree/81bf79fd9aa75305e55abe2f7e9ef3f60624a3a1/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters

Without the custom converter, deserialization puts a `JsonElement` in each property.

> The [unit tests folder](https://github.com/dotnet/corefx/blob/81bf79fd9aa75305e55abe2f7e9ef3f60624a3a1/src/System.Text.Json/tests/Serialization/) in the `System.Text.Json.Serialization` namespace has more examples of custom converters that handle deserialization to Object properties.

Reviewer 2: This link is 404. Needs to be dotnet/runtime not dotnet/corefx.

Any other links that have corefx in them?

```suggestion
The [unit tests folder](https://github.com/dotnet/runtime/blob/81bf79fd9aa75305e55abe2f7e9ef3f60624a3a1/src/System.Text.Json/tests/Serialization/) in the `System.Text.Json.Serialization` namespace has more examples of custom converters that handle deserialization to Object properties.
```

}
```

> The [unit tests folder](https://github.com/dotnet/corefx/blob/81bf79fd9aa75305e55abe2f7e9ef3f60624a3a1/src/System.Text.Json/tests/Serialization/) in the `System.Text.Json.Serialization` namespace has more examples of custom converters that handle non-string-key dictionaries.

Reviewer 2: ```suggestion
The [unit tests folder](https://github.com/dotnet/runtime/blob/81bf79fd9aa75305e55abe2f7e9ef3f60624a3a1/src/System.Text.Json/tests/Serialization/) in the `System.Text.Json.Serialization` namespace has more examples of custom converters that handle non-string-key dictionaries.
```

### Support polymorphic deserialization

> Built-in features provide a limited range of [polymorphic serialization](system-text-json-how-to.md#serialize-properties-of-derived-classes) but not deserialization. Deserialization requires a custom converter.

Reviewer 2: nit: Kinda reads like deserialization doesn't have limited range of support (i.e. has full support).

```suggestion
Built-in features provide a limited range of [polymorphic serialization](system-text-json-how-to.md#serialize-properties-of-derived-classes) but provide no support for deserialization. Deserialization requires a custom converter.
```

}
```

The [unit tests folder](https://github.com/dotnet/corefx/blob/f77914d4c9045a01b3a91b63c28805f24850c274/src/System.Text.Json/tests/Serialization/) in the `System.Text.Json.Serialization` namespace has more examples of custom converters that handle non-string-key dictionaries.

> ### Support polymorphic deserialization

Reviewer 2: This section needs to be corrected a bit now. Specifically, the following opening isn't accurate:

> Polymorphic serialization doesn't require a custom converter, but deserialization does require a custom converter.

## Other custom converter samples

> The [Migrate from Newtonsoft.Json to System.Text.Json](system-text-json-migrate-from-newtonsoft-how-to.md) article contains samples of custom converters.

Reviewer 2: This comment/link is outdated now.

Author: It does have examples of other converters, just not the three that are in this doc.

Reviewer 2: Like what, and are they useful/different enough to provide value? Should the link point to the specific converter snippet?

Also, maybe re-word as:
```suggestion
The [Migrate from Newtonsoft.Json to System.Text.Json](system-text-json-migrate-from-newtonsoft-how-to.md) article also contains some samples of custom converters.
```
